File: RumoPragas/Services/SupabaseService.swift
import Foundation

nonisolated final class SupabaseService: Sendable {
    static let shared = SupabaseService()

    private let supabaseURL: String
    private let supabaseKey: String

    private init() {
        self.supabaseURL = Config.EXPO_PUBLIC_SUPABASE_URL
        self.supabaseKey = Config.EXPO_PUBLIC_SUPABASE_ANON_KEY
    }

    private var baseURL: String { supabaseURL }

    private func makeRequest(
        path: String,
        method: String = "GET",
        body: Data? = nil,
        token: String? = nil,
        additionalHeaders: [String: String] = [:]
    ) -> URLRequest? {
        guard let url = URL(string: "\(baseURL)\(path)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(supabaseKey, forHTTPHeaderField: "apikey")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        } else {
            request.setValue("Bearer \(supabaseKey)", forHTTPHeaderField: "Authorization")
        }
        for (key, value) in additionalHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }
        request.httpBody = body
        return request
    }

    func signUp(email: String, password: String, fullName: String) async throws -> AuthResponse {
        let payload: [String: Any] = [
            "email": email,
            "password": password,
            "data": ["full_name": fullName]
        ]
        let body = try JSONSerialization.data(withJSONObject: payload)
        guard let request = makeRequest(path: "/auth/v1/signup", method: "POST", body: body) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw parseAuthError(data: data)
        }
        return try JSONDecoder().decode(AuthResponse.self, from: data)
    }

    func signIn(email: String, password: String) async throws -> AuthResponse {
        let payload = ["email": email, "password": password]
        let body = try JSONEncoder().encode(payload)
        guard let request = makeRequest(
            path: "/auth/v1/token?grant_type=password",
            method: "POST",
            body: body
        ) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw parseAuthError(data: data)
        }
        return try JSONDecoder().decode(AuthResponse.self, from: data)
    }

    func resetPassword(email: String) async throws {
        let payload = ["email": email]
        let body = try JSONEncoder().encode(payload)
        guard let request = makeRequest(path: "/auth/v1/recover", method: "POST", body: body) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw parseAuthError(data: data)
        }
    }

    func signOut(token: String) async throws {
        guard let request = makeRequest(path: "/auth/v1/logout", method: "POST", token: token) else {
            return
        }
        let _ = try await URLSession.shared.data(for: request)
    }

    func refreshToken(_ refreshToken: String) async throws -> AuthResponse {
        let payload = ["refresh_token": refreshToken]
        let body = try JSONEncoder().encode(payload)
        guard let request = makeRequest(
            path: "/auth/v1/token?grant_type=refresh_token",
            method: "POST",
            body: body
        ) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw parseAuthError(data: data)
        }
        return try JSONDecoder().decode(AuthResponse.self, from: data)
    }

    func getUser(token: String) async throws -> SupabaseUser {
        guard let request = makeRequest(path: "/auth/v1/user", token: token) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw parseAuthError(data: data)
        }
        return try JSONDecoder().decode(SupabaseUser.self, from: data)
    }

    private func parseAuthError(data: Data) -> APIError {
        if let errorObj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let msg = errorObj["error_description"] as? String ?? errorObj["msg"] as? String ?? errorObj["error"] as? String ?? errorObj["message"] as? String {
                return .serverError(msg)
            }
        }
        return .authFailed
    }

    func fetchDiagnoses(token: String, userId: String, limit: Int = 50) async throws -> [DiagnosisResult] {
        let encodedUserId = userId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userId
        guard let request = makeRequest(
            path: "/rest/v1/pragas_diagnoses?user_id=eq.\(encodedUserId)&order=created_at.desc&limit=\(limit)",
            token: token,
            additionalHeaders: ["Prefer": "return=representation"]
        ) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw APIError.networkError
        }
        return try JSONDecoder().decode([DiagnosisResult].self, from: data)
    }

    func fetchDiagnosisCount(token: String, userId: String) async throws -> Int {
        let encodedUserId = userId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userId
        guard var request = makeRequest(
            path: "/rest/v1/pragas_diagnoses?user_id=eq.\(encodedUserId)&select=id",
            token: token,
            additionalHeaders: ["Prefer": "count=exact"]
        ) else {
            throw APIError.invalidURL
        }
        request.httpMethod = "HEAD"
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw APIError.networkError
        }
        if let range = http.value(forHTTPHeaderField: "Content-Range"),
           let total = range.split(separator: "/").last,
           let count = Int(total) {
            return count
        }
        return 0
    }

    func deleteDiagnosis(token: String, id: String) async throws {
        let encodedId = id.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? id
        guard let request = makeRequest(
            path: "/rest/v1/pragas_diagnoses?id=eq.\(encodedId)",
            method: "DELETE",
            token: token
        ) else {
            throw APIError.invalidURL
        }
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw APIError.networkError
        }
    }

    func updateProfile(token: String, userId: String, profile: [String: Any]) async throws {
        let encodedUserId = userId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userId
        let body = try JSONSerialization.data(withJSONObject: profile)
        guard let request = makeRequest(
            path: "/rest/v1/pragas_profiles?id=eq.\(encodedUserId)",
            method: "PATCH",
            body: body,
            token: token,
            additionalHeaders: ["Prefer": "return=representation"]
        ) else {
            throw APIError.invalidURL
        }
        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw APIError.networkError
        }
    }

    func fetchProfile(token: String, userId: String) async throws -> UserProfile? {
        let encodedUserId = userId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userId
        guard let request = makeRequest(
            path: "/rest/v1/pragas_profiles?id=eq.\(encodedUserId)&limit=1",
            token: token
        ) else {
            throw APIError.invalidURL
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw APIError.networkError
        }
        let profiles = try JSONDecoder().decode([UserProfile].self, from: data)
        return profiles.first
    }

    func callEdgeFunction(
        name: String,
        body: Data?,
        token: String
    ) async throws -> Data {
        guard let url = URL(string: "\(baseURL)/functions/v1/\(name)") else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(supabaseKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = body
        request.timeoutInterval = 180

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch let urlError as URLError {
            switch urlError.code {
            case .timedOut:
                throw APIError.serverError("Tempo esgotado. A análise demorou demais. Tente com uma imagem menor.")
            case .notConnectedToInternet, .networkConnectionLost:
                throw APIError.serverError("Sem conexão com a internet. Verifique sua rede.")
            default:
                throw APIError.networkError
            }
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.networkError
        }

        #if DEBUG
        print("[EdgeFunction] \(name) -> HTTP \(http.statusCode), \(data.count) bytes")
        #endif

        if (200...299).contains(http.statusCode) {
            return data
        }

        #if DEBUG
        let rawBody = String(data: data, encoding: .utf8) ?? "(empty)"
        print("[EdgeFunction] Error body: \(rawBody.prefix(1000))")
        #endif

        if let errorObj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let errMsg = errorObj["error"] as? String {
                throw APIError.serverError(errMsg)
            }
            if let msg = errorObj["message"] as? String {
                throw APIError.serverError(msg)
            }
            if let msg = errorObj["msg"] as? String {
                throw APIError.serverError(msg)
            }
        }

        if http.statusCode == 404 {
            throw APIError.serverError("Função '\(name)' não encontrada no servidor. Verifique se a Edge Function está publicada.")
        }
        if http.statusCode == 401 || http.statusCode == 403 {
            throw APIError.serverError("Sessão expirada. Faça login novamente.")
        }
        if http.statusCode == 500 {
            throw APIError.serverError("Erro interno do servidor. Tente novamente em alguns instantes.")
        }
        throw APIError.serverError("Erro do servidor (HTTP \(http.statusCode))")
    }
}

nonisolated struct EdgeFunctionError: Codable, Sendable {
    let error: String
}

nonisolated struct AuthResponse: Codable, Sendable {
    let accessToken: String?
    let tokenType: String?
    let refreshToken: String?
    let user: SupabaseUser?

    nonisolated enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case tokenType = "token_type"
        case refreshToken = "refresh_token"
        case user
    }
}

nonisolated struct SupabaseUser: Codable, Sendable {
    let id: String
    let email: String?
    let userMetadata: UserMetadata?

    nonisolated enum CodingKeys: String, CodingKey {
        case id
        case email
        case userMetadata = "user_metadata"
    }
}

nonisolated struct UserMetadata: Codable, Sendable {
    let fullName: String?

    nonisolated enum CodingKeys: String, CodingKey {
        case fullName = "full_name"
    }
}

nonisolated enum APIError: LocalizedError, Sendable {
    case invalidURL
    case networkError
    case authFailed
    case decodingError
    case serverError(String)
    case subscriptionRequired

    var errorDescription: String? {
        switch self {
        case .invalidURL: "URL inválida"
        case .networkError: "Erro de conexão. Verifique sua internet."
        case .authFailed: "Falha na autenticação"
        case .decodingError: "Erro ao processar resposta"
        case .serverError(let msg): msg
        case .subscriptionRequired: "Assinatura necessária para usar o diagnóstico por IA."
        }
    }
}

File: web-android/src/services/supabaseService.ts
import { Config } from '../utils/config';
import {
  AuthResponse,
  SupabaseUser,
  DiagnosisResult,
  UserProfile,
} from '../types';

// ─── Error class ─────────────────────────────────────────────────────────────

export class APIError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'APIError';
  }

  static invalidURL = new APIError('URL inválida');
  static networkError = new APIError('Erro de conexão. Verifique sua internet.');
  static authFailed = new APIError('Falha na autenticação');
  static decodingError = new APIError('Erro ao processar resposta');
  static subscriptionRequired = new APIError('Assinatura necessária para usar o diagnóstico por IA.');
}

// ─── Helper ──────────────────────────────────────────────────────────────────

function parseAuthError(data: any): APIError {
  if (data && typeof data === 'object') {
    const msg =
      data.error_description ?? data.msg ?? data.error ?? data.message;
    if (typeof msg === 'string') {
      return new APIError(msg);
    }
  }
  return new APIError('Falha na autenticação');
}

async function makeRequest(
  path: string,
  options: {
    method?: string;
    body?: any;
    token?: string;
    additionalHeaders?: Record<string, string>;
    timeoutMs?: number;
  } = {},
): Promise<Response> {
  const {
    method = 'GET',
    body,
    token,
    additionalHeaders = {},
    timeoutMs,
  } = options;

  const url = `${Config.supabaseURL}${path}`;

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    apikey: Config.supabaseAnonKey,
    Authorization: token
      ? `Bearer ${token}`
      : `Bearer ${Config.supabaseAnonKey}`,
    ...additionalHeaders,
  };

  const fetchOptions: RequestInit = { method, headers };
  if (body !== undefined) {
    fetchOptions.body = typeof body === 'string' ? body : JSON.stringify(body);
  }

  if (timeoutMs) {
    const controller = new AbortController();
    fetchOptions.signal = controller.signal;
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const res = await fetch(url, fetchOptions);
      clearTimeout(timer);
      return res;
    } catch (err: any) {
      clearTimeout(timer);
      if (err.name === 'AbortError') {
        throw new APIError(
          'Tempo esgotado. A análise demorou demais. Tente com uma imagem menor.',
        );
      }
      throw new APIError('Erro de conexão. Verifique sua internet.');
    }
  }

  return fetch(url, fetchOptions);
}

// ─── Service ─────────────────────────────────────────────────────────────────

export const SupabaseService = {
  // ── Auth ──────────────────────────────────────────────────────────────────

  async signUp(
    email: string,
    password: string,
    fullName: string,
  ): Promise<AuthResponse> {
    const res = await makeRequest('/auth/v1/signup', {
      method: 'POST',
      body: { email, password, data: { full_name: fullName } },
    });
    const data = await res.json();
    if (!res.ok) throw parseAuthError(data);
    return data as AuthResponse;
  },

  async signIn(email: string, password: string): Promise<AuthResponse> {
    const res = await makeRequest('/auth/v1/token?grant_type=password', {
      method: 'POST',
      body: { email, password },
    });
    const data = await res.json();
    if (!res.ok) throw parseAuthError(data);
    return data as AuthResponse;
  },

  async resetPassword(email: string): Promise<void> {
    const res = await makeRequest('/auth/v1/recover', {
      method: 'POST',
      body: { email },
    });
    if (!res.ok) {
      const data = await res.json();
      throw parseAuthError(data);
    }
  },

  async signOut(token: string): Promise<void> {
    await makeRequest('/auth/v1/logout', {
      method: 'POST',
      token,
    });
  },

  async refreshToken(refreshToken: string): Promise<AuthResponse> {
    const res = await makeRequest('/auth/v1/token?grant_type=refresh_token', {
      method: 'POST',
      body: { refresh_token: refreshToken },
    });
    const data = await res.json();
    if (!res.ok) throw parseAuthError(data);
    return data as AuthResponse;
  },

  async getUser(token: string): Promise<SupabaseUser> {
    const res = await makeRequest('/auth/v1/user', { token });
    const data = await res.json();
    if (!res.ok) throw parseAuthError(data);
    return data as SupabaseUser;
  },

  // ── Data ──────────────────────────────────────────────────────────────────

  async fetchDiagnoses(
    token: string,
    userId: string,
    limit: number = 200,
  ): Promise<DiagnosisResult[]> {
    const res = await makeRequest(
      `/rest/v1/pragas_diagnoses?user_id=eq.${userId}&order=created_at.desc&limit=${limit}`,
      {
        token,
        additionalHeaders: { Prefer: 'return=representation' },
      },
    );
    if (!res.ok) throw new APIError('Erro de conexão. Verifique sua internet.');
    return (await res.json()) as DiagnosisResult[];
  },

  async countDiagnoses(token: string, userId: string): Promise<number> {
    const res = await makeRequest(
      `/rest/v1/pragas_diagnoses?user_id=eq.${userId}&select=id`,
      {
        token,
        additionalHeaders: {
          Prefer: 'count=exact',
          'Range-Unit': 'items',
          Range: '0-0',
        },
      },
    );
    if (!res.ok) throw new APIError('Erro de conexão. Verifique sua internet.');
    const rangeHeader = res.headers.get('content-range');
    if (rangeHeader) {
      const parts = rangeHeader.split('/');
      const total = parts[parts.length - 1];
      const count = parseInt(total, 10);
      if (!isNaN(count)) return count;
    }
    return 0;
  },

  async saveDiagnosis(
    token: string,
    diagnosis: Record<string, any>,
  ): Promise<DiagnosisResult> {
    const res = await makeRequest('/rest/v1/pragas_diagnoses', {
      method: 'POST',
      body: diagnosis,
      token,
      additionalHeaders: { Prefer: 'return=representation' },
    });
    if (!res.ok) throw new APIError('Erro ao salvar diagnóstico.');
    const rows = await res.json();
    return (Array.isArray(rows) ? rows[0] : rows) as DiagnosisResult;
  },

  async deleteDiagnosis(token: string, id: string): Promise<void> {
    const res = await makeRequest(`/rest/v1/pragas_diagnoses?id=eq.${id}`, {
      method: 'DELETE',
      token,
    });
    if (!res.ok) throw new APIError('Erro de conexão. Verifique sua internet.');
  },

  async updateProfile(
    token: string,
    userId: string,
    profile: Record<string, any>,
  ): Promise<void> {
    // Use upsert so new users get a profile row created automatically
    const res = await makeRequest('/rest/v1/pragas_profiles', {
      method: 'POST',
      body: { id: userId, ...profile },
      token,
      additionalHeaders: {
        Prefer: 'return=representation,resolution=merge-duplicates',
      },
    });
    if (!res.ok) throw new APIError('Erro de conexão. Verifique sua internet.');
  },

  async fetchProfile(
    token: string,
    userId: string,
  ): Promise<UserProfile | null> {
    const res = await makeRequest(
      `/rest/v1/pragas_profiles?id=eq.${userId}&limit=1`,
      { token },
    );
    if (!res.ok) throw new APIError('Erro de conexão. Verifique sua internet.');
    const profiles = (await res.json()) as UserProfile[];
    return profiles.length > 0 ? profiles[0] : null;
  },

  // ── Edge Functions ────────────────────────────────────────────────────────

  async callEdgeFunction(
    name: string,
    body: any | null,
    token: string,
  ): Promise<any> {
    const url = `${Config.supabaseURL}/functions/v1/${name}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 180_000);

    let res: Response;
    try {
      res = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          apikey: Config.supabaseAnonKey,
          Authorization: `Bearer ${token}`,
        },
        body: body != null ? JSON.stringify(body) : undefined,
        signal: controller.signal,
      });
      clearTimeout(timer);
    } catch (err: any) {
      clearTimeout(timer);
      if (err.name === 'AbortError') {
        throw new APIError(
          'Tempo esgotado. A análise demorou demais. Tente com uma imagem menor.',
        );
      }
      throw new APIError('Erro de conexão. Verifique sua internet.');
    }

    const data = await res.json().catch(() => null);

    console.log(
      `[EdgeFunction] ${name} -> HTTP ${res.status}, ${JSON.stringify(data)?.length ?? 0} bytes`,
    );

    if (res.ok) return data;

    const rawBody = JSON.stringify(data) ?? '(empty)';
    console.log(`[EdgeFunction] Error body: ${rawBody.substring(0, 1000)}`);

    if (data && typeof data === 'object') {
      if (typeof data.error === 'string') throw new APIError(data.error);
      if (typeof data.message === 'string') throw new APIError(data.message);
      if (typeof data.msg === 'string') throw new APIError(data.msg);
    }

    if (res.status === 404) {
      throw new APIError(
        `Função '${name}' não encontrada no servidor. Verifique se a Edge Function está publicada.`,
      );
    }
    if (res.status === 401 || res.status === 403) {
      throw new APIError('Sessão expirada. Faça login novamente.');
    }
    if (res.status === 500) {
      throw new APIError(
        'Erro interno do servidor. Tente novamente em alguns instantes.',
      );
    }
    throw new APIError(`Erro do servidor (HTTP ${res.status})`);
  },
};
